feat: add the key or index as an argument of the equality comparator

g-traub opened the pull request:
Fixes #56

## Why?
Make createCustomEqual more flexible by passing it the key or index of the values. To make it possible to ignore properties for example.

## How?
- Pass the key for objects and Maps and index for arrays to the `isEqual` function.
- Change the type definition of `EqualityComparator` in utils.ts

planttheidea commented:
I won't have the opportunity to review it this week, but my initial impression is support of this. That said, I'd rather align with the same contract that `lodash` has, both for portability and completeness (it has all the things I can imagine for a comparison, might as well go big).

It would require a breaking change, but I doubt it is one that will impact many consumers as most don't use the custom comparator option.

g-traub commented:
I'm glad you approve of the idea. What do you mean by "I'd rather align with the same contract that lodash has" ? Do you mean adding more elements to the customizer function like `parent`, `otherParent`, `stack` ? If so, what about the meta parameter currently used ?
The lodash isEqualWith customizer function signature :
```ts
type IsEqualCustomizer = (value: any, other: any, indexOrKey: PropertyName | undefined, parent: any, otherParent: any, stack: any) => boolean | undefined;
```
The code changes I suggested are minimal and don't require a breaking change, but if you think more changes are necessary, I'll be happy to look into it and update this PR.

planttheidea commented:
Indeed, I was referring to matching that contract. The `stack` parameter is conceptually the same as the `meta` parameter, so my thinking is:

```ts
type EqualityComparator = (a: any, b: any, indexOrKey: any, parentA: any, parentB: any, meta?: any) => boolean;
```

What do you think?

g-traub commented:
I think it's a good idea! 
Regarding the meta parameter :
- Do you want to turn `meta` into an `stack`-like object that will always track traversed objects, as it's done now for `circularDeepEqual` and `circularShallowEqual`? (And remove it from the parameters in the function returned by `createCustomEqual` since it will be used only internally?)
- Or do you want to keep it as it is now ?

planttheidea commented:
It's rather keep it as it is now. Speed gains come from avoiding that stack except for when needed (`circularDeepEqual` vs `deepEqual`), and it is also more flexible for custom comparators.

I'd also prefer to keep `meta` last in the order of parameters, instead of just appending the new ones. The ordering I mention in the previous comment aligns with lodash, but more importantly it prioritizes the cake of information better. `meta` is basically a catch-all if all the preceding parameters don't suffice, and making developers list an unused parameter frequently is unnecessary friction.

g-traub commented:
Ok I understand. I'll change the order !

planttheidea commented on README.md:
````diff
@@ -157,24 +157,25 @@ The signature is as follows:
 
 ```typescript
 type EqualityComparator = (a: any, b: any, meta?: any) => boolean;
+type ExtendedEqualityComparator = (a: any, b: any, indexOrKey?: any, parentA?: any, parentB?: any, meta?: any) => boolean;
````
Rather than have two types of equality comparators, I believe a consistent one should be used. It would involve a small change internally, but it makes the custom comparator use-case more flexible.

g-traub replied:
If you mean a single type for every comparator, I think it would be confusing. The comparator returned by `createCustomEqual` would have indexOrKey, parentA, parentB, which are not supposed to be user provided. The example in the docs would become :
```ts
const isDeepEqualOrFooMatchesMeta = createCustomEqual(
  (deepEqual) => (objectA, objectB, indexOrKey, parentA, parentB, meta) =>
    objectA.foo === meta ||
    objectB.foo === meta ||
    deepEqual(objectA, objectB, indexOrKey, parentA, parentB, meta)
);
isDeepEqualOrFooMatchesMeta({ foo: 'bar' }, { foo: 'baz' },  undefined, undefined, undefined, 'bar')
```

Another way we could do it is to have all internal comparators a single type and only a different type for the "external" comparator (the return of `createComparator`) ? It would allow us to simplify the `EqualityComparatorCreator` type and thus the docs :
```ts
export type EqualityComparatorCreator = (fn: EqualityComparator) => EqualityComparator;
```
But it would mean having the user type `indexOrKey`, `parentA`, `parentB` in the `deepEqual` call if he uses meta, when it's not necessary (since those properties are only for the custom comparator and not used in `comparator`) :
```ts
const isDeepEqualOrFooMatchesMeta = createCustomEqual(
  (deepEqual) => (objectA, objectB, indexOrKey, parentA, parentB, meta) =>
    deepEqual(objectA, objectB, indexOrKey, parentA, parentB, meta), // <-- Here
);
```
Plus, I think it will be confusing that the `deepEqual` passed as an argument doesn't take the same arguments as `isDeepEqualOrFooMatchesMeta`.
And it would mean still having two types : one for the methods exposed by the package and one for the internal logic.

I'm not sure about this... I think the current way it's done is the less confusing one. 
Maybe I didn't understand what you had in mind.

planttheidea replied:
Ah yes, as you lay out the usage what you describe makes sense, in terms of the two separate contracts. Based on this, my only recommendation then would be to rename the "extended" comparator to instead be the "internal" comparator, to better explain the different uses.

g-traub replied:
I agree, it' difficult to find a good name for it...
"Extended" only conveys that there is something more to it than the other one, but not what it's used for.
"Internal" is better since it conveys where it's used. Even though the user interact with, it's as a returned function. The user provides the comparator that will be used internally. I think it could be an ok name.
```ts
type EqualityComparatorCreator = (EqualityComparator) => InternalEqualityComparator;
```
I think that having `IsEqual` and `EqualityComparator` will be confusing. It is the same thing expressed differently. Without context, choosing between the two for typing `deepEqual` would be impossible. Plus, `isEqual` is the name of many other libraries export (like lodash).
I think "internal" is the one that provides the most information, even though it's not perfect.

planttheidea commented on src/utils.ts:
```diff
@@ -294,7 +310,7 @@ export function areSetsEqual(
         b.forEach((bValue) => {
           if (!isValueEqual) {
-            isValueEqual = isEqual(aValue, bValue, meta);
+            isValueEqual = isEqual(aValue, bValue, undefined, a, b, meta);
```
Hmm ... I realize the issue here, but having an `undefined` key feels unusual. In `Set.forEach()`, the `key` is the same as the `value`, and since `b` is the driver here perhaps pass through from the iteration:

```ts
b.forEach((bValue, bKey) => {
  if (!isValueEqual) {
    isValueEqual = isEqual(aValue, bValue, bKey, a, b, meta);
```

Another possibility would be to keep a local counter, and pass it through:

```ts
let iterationIndex = 0;

b.forEach((bValue) => {
  if (!isValueEqual) {
    isValueEqual = isEqual(aValue, bValue, iterationIndex++, a, b, meta);
```

This may be helpful for order-based comparisons in a custom comparator. Thoughts?

g-traub replied:
I think your second proposition with an `iterationIndex` is a good idea. The only thing is it could break if parentA or B is access with index (parentA[index]), but since the value is already provided, why would it be ? We could provide the parents as arrays, but I think the performance drop that it would cause is not worth it.

planttheidea replied:
Yup, I understand the dilemma, but doesn't the same issue exist with `undefined`? `parentA[undefined]` would break in the same way.

g-traub replied:
Right, I didn't think of `parentA.get(undefined)`.
I'll add the `iterationIndex` !

planttheidea commented on src/utils.ts:
```diff
@@ -178,8 +194,8 @@ export function areMapsEqual(
         b.forEach((bValue, bKey) => {
-          if (!isValueEqual && isEqual(aKey, bKey, meta)) {
-            isValueEqual = isEqual(aValue, bValue, meta);
+          if (!isValueEqual && isEqual(aKey, bKey, undefined, a, b, meta)) {
```
I mention it in more detail below related to `Set` comparisons, but `key` being `undefined` here makes sense from an implementation perspective, but may not from a consumption perspective. The idea I had there related to an "iteration index" may be applicable here as well.

g-traub replied:
Here, I think we shouldn't use a custom comparator and pass these values. This comparison only makes sense internally to make sure we compare the right two Map values. When I wrote the test for the new params I forgot this case at first, and I think users of the library don't know keys from Maps will show up as values in the custom comparator when using `createCustomEqual`. It could even break the comparison for some extreme cases.

From the specs for Map:

> Distinct key values are discriminated using the SameValueZero comparison algorithm.

So we could use sameValueZero comparison, or use `.has` and `.get` instead of the second loop. (This is the approach taken by fast-deep-equal). 
Or we could keep deep comparison of keys but use the basic `compare`.
What do you think ?

planttheidea replied:
The reason for this internal comparison is to allow value-based comparisons. Your reference to the spec is correct, however an architectural decision was "deep equality is not exclusive to values". Example:

```ts
const mapA = new Map([[{ foo: 'bar'}, { bar: 'baz' }]]);
const mapB = new Map([[{ foo: 'bar'}, { bar: 'baz' }]]);
```

If we used SMZ for comparison, these would never be equal, even though the map contents (key + value) are identical.

In regards to the second point of using the basic `compare` instead of the provided `isEqual`, I went back and forth on this but eventually came to the decision that creating different comparisons for key vs value left a gap in capabilities. Creating a custom equality comparator is a rare use-case, so instead of limit functionality (what if a custom `key` comparison was desired?) I decided to leave the implementation details of the comparison to the consumer. I try to follow a general guideline of "common use-cases are simple, advanced use-cases are understandable, extreme edge-cases are possible".

g-traub replied:
Ok, I understand, it makes sense. We could maybe add a line about it in the docs ?

planttheidea replied:
Good call, it definitely should be mentioned as it is not obvious. Prior to publish of this change I'll update the docs to provide that nuance.

planttheidea reviewed (changes requested):
The main request for changes is unifying the equality comparators. I've solicited feedback on `undefined` keys as well, because there is some ambiguity there.

planttheidea commented:
@g-traub - apologies for the churn, but I wanted to release a patch version with a fix for #58 before this breaking change for consumers using `^2.x.x`. Hopefully the conflict resolutions aren't too painful.

planttheidea reviewed (approved):
Changes look good. :+1: Thanks for the contribution!

g-traub commented:
Resolved the conflicts and fixed 2 issues I previously missed.
Thank you for your patience, help and feedback !

planttheidea commented on src/utils.ts:
```diff
@@ -183,7 +199,7 @@ export function areMapsEqual(
           if (!hasMatch && !matchedIndices[matchIndex]) {
             hasMatch =
-              isEqual(aKey, bKey, meta) && isEqual(aValue, bValue, meta);
+            isEqual(aKey, bKey, matchIndex, a, b, meta) && isEqual(aValue, bValue, aKey, a, b, meta);
```
I just noticed `aKey` being used here instead of `matchIndex`, which creates a difference from how `Set` operates and also could be confusing since `b.get(aKey)` may return `false` even though the equality check could return `true`.

g-traub replied:
You're right, it's not straightforward... This could return `undefined` if the key is not a primitive, and we're not using shallowEqual. It is inconsistent.
I think we get more information by passing `aKey` but the tradeoff is that inconsistency.
Having the same behaviour of passing the `matchIndex` for `Set` and `Map` would be confusing in my opinion, since Maps are supposed to be keyed.
A way to have it consistent between `Map` and `Set` would be to use arrays (converting) to compare them. This way it would be accessible with parent[index].

planttheidea replied:
## Option 1: always using `matchIndex`

An important distinction is that the definition of equality for this library differs from the spec for `Map` / `Set`. Example:

```ts
const mapA = new Map([[{ foo: 'bar' }, 'baz']]);
const mapB = new Map([[{ foo: 'bar' }, 'baz']]);

deepEqual(mapA, mapB); // true
```
This means that for `Map` specifically, providing only one or the other key would provide inaccurate information; `b.get(aKey)` with the above example would be invalid. This means confusion will always exist, just shifted dependency on the decision here.

Accepting that confusion will always exist, generally implementation consistency coupled with documentation is the chosen direction. With that in mind, something like this in the README section on `createCustomEqual`:

```md
**NOTE**: `Map` and `Set` implementations may have different key references and still be considered equal depending on your needs. Because order may be an aspect of the decision, the iteration index is provided as the `key` in the custom comparator. If you need to do a lookup in either object based on this index, it is recommended to use native methods like `.keys()` or `.values()` in combination with the index.
```

We can even provide an example, something like:

```js
function customComparator(aValue, bValue, iterationIndex, a, b, meta) {
  const aKey = Array.from(a.keys())[iterationIndex];
```

## Option 2: Embrace the difference in keys

We could optionally tweak the contract to support the different multiple keys:

```ts
export type InternalEqualityComparator = (
  objectA: any,
  objectB: any,
  indexOrKeyA: any,
  indexOrKeyB: any,
  parentA: any,
  parentB: any,
  meta: any,
) => boolean;
```

This would be duplicative in array / object operations:

```ts
// arrays
if (!isEqual(a[index], b[index], index, index, a, b, meta)) {
// objects
if (!hasOwnProperty(b, key) || !isEqual(a[key], b[key], key, key, a, b, meta)) {
```

But would allow full pass-through in `Map` / `Set` operations:

```ts
// maps
hasMatch = isEqual(aKey, bKey, matchIndex, matchIndex, a, b, meta) && isEqual(aValue, bValue, aKey, bKey, a, b, meta);
// sets
hasMatch = isEqual(aValue, bValue, aValue, bValue, a, b, meta);
```

At this point the only inconsistency would be the passing of `matchIndex` for doing `Map`-based key equality comparisons, however again documentation could explain this:

```md
**NOTE**: `Map` implementations compare equality for both keys and value. When using a custom comparator and comparing equality of the keys, the iteration index is provided as both `aKey` and `bKey` to help use-cases where ordering of keys matters to equality.
```

For all scenarios other than `Map` the multiple keys would not be much help, but the only cost would be a "throwaway" function parameter.

g-traub replied:
> generally implementation consistency coupled with documentation is the chosen direction

I completely agree !

Using `matchIndex` has a problem : the index is only accurate for `b`. Sets and Maps don't take into account the order of elements when checking for equality and the index is incremented in the second loop, looping over b values. 
``` ts
  const setA = new Set(['x', 'y']);
  const setB = new Set(['y', 'x']);

  const createIsEqual: EqualityComparatorCreator =
    (deepEqual) => (a: any,
      b: any,
      indexOrKey?: any,
      parentA?: any,
      parentB?: any,
      meta?: any) => {
      expect(Array.from(parentB)[indexOrKey]).toBe(b);
      expect(Array.from(parentA)[indexOrKey]).toBe(a);

      return deepEqual(a, b, meta);
    };
  const comparator = createComparator(createIsEqual);
  comparator(setA, setB);
```
This test fails on the first `isEqual` call internally. The value passed is `'x'`, but at the index passed `1`, for `a`, the value is `'y'`

So I think we should go with option 2. Having 2 index/key params enables us to add another index that increments in the `a` loop and give two accurate indexes for doing Map-based key equality comparisons. What do you think ?

planttheidea replied:
Yup, agreed on all fronts!

planttheidea reviewed (changes requested):
One last thing to discuss, otherwise LGTM. It is less a request for changes, and more a request for discussion before resolution.

planttheidea commented:
> Resolved the conflicts and fixed 2 issues I previously missed. Thank you for your patience, help and feedback !

No worries! And I appreciate your patience, which I put to the test :sweat_smile: by having one more thing to discuss before merge.

g-traub commented:
What's your opinion about `areMapsEqual` using `aKey` ? Do you think we should change it ?

planttheidea commented:
Apologies for the delay; it was a bit of a long winter for me. I'll try to get back to this over the weekend.

g-traub commented:
No worries, there's no rush 🙂

g-traub commented:
Done :+1: 
- I used option 2, iteration index for Map keys and values for Sets.
- Added note taken from your comment to the `README.md`

### Other change
I realized while doing these changes that no typescript error popped up since all parameters were optional... But any missing parameter in `InternalEqualityComparator` could break a custom comparator, so I changed the parameters to required.
It was necessary to change the `createCircularEqualCreator` test since it calls an `InternalEqualityComparator` directly. I added `undefined` for the other params which isn't great but at least the change to required params give us more confidence that `InternalEqualityComparator` is used correctly.

### Merging
Since this branch has 14 commits, I recommend using "squash and merge" to keep the history clean.
